File: 力扣习题/204计数质数/problems.py
import math


class Solution:
    def countPrimes(self, n: int) -> int:
        '''
        j并不是从2*i开始，而是i*i开始，i也不需要到n，到sqrt(n)即可
        '''
        if n == 0 or n == 1:
            return 0
        p = [1] * n
        i = 2
        while i * i < n:
            if p[i]:
                for j in range(i * i, n, i):
                    p[j] = 0
            i += 1
        return p[2:n].count(1)

s = Solution()

print(s.countPrimes(10))

File: 力扣习题/204计数质数/problems.ts
function countPrimes(n: number): number {
    const p: number[] = new Array(n);
    p.fill(1);
    p[0] = 0;
    p[1] = 0;
    for (let i = 2; i * i < n; i++) {
        if (p[i]) {
            for (let j = i * i; j < n; j += i) {
                p[j] = 0;
            }
        }
    }
    let c = 0;
    for (let i = 2; i < n; i++) {
        if (p[i]) {
            c++;
        }
    }
    return c;
};